src/sentry/rules/processing/delayed_processing.py

```diff
@@ -1,36 +1,273 @@
+    rules_to_fire = defaultdict(set)
+    for alert_rule, slow_conditions in rule_to_slow_conditions.items():
+        for slow_condition in slow_conditions:
+            condition_id = slow_condition.get("id")
+            target_value = int(slow_condition.get("value"))
+            results = condition_to_results[condition_id]
+            for group_id in rules_to_groups[alert_rule.id]:
+                if results[group_id] > target_value:
+                    rules_to_fire[alert_rule].add(group_id)
+    return rules_to_fire
+
+
 @redis_buffer_registry.add_handler(BufferHookEvent.FLUSH)
 def process_delayed_alert_conditions(buffer: RedisBuffer) -> None:
     with metrics.timer("delayed_processing.process_all_conditions.duration"):
         project_ids = buffer.get_set(PROJECT_ID_BUFFER_LIST_KEY)
 
         for project in RangeQuerySetWrapper(Project.objects.filter(id__in=project_ids)):
-            rulegroup_event_mapping = buffer.get_hash(
-                model=Project, field={"project_id": project.id}
-            )
-
             with metrics.timer("delayed_processing.process_project.duration"):
-                safe_execute(
-                    apply_delayed,
-                    project,
-                    rulegroup_event_mapping,
-                    _with_transaction=False,
-                )
+                apply_delayed.delay(project=project, buffer=buffer)
+
+
+@instrumented_task(
+    name="sentry.delayed_processing.tasks.apply_delayed",
+    queue="dynamicsampling",
+    default_retry_delay=5,
+    max_retries=5,
+    soft_time_limit=50,
+    time_limit=60,  # 1 minute
+    silo_mode=SiloMode.REGION,
+)
+def apply_delayed(project: Project, buffer: RedisBuffer) -> None:
+    """
+    Grab rules, groups, and events from the Redis buffer, evaluate the "slow" conditions in a bulk snuba query, and fire them if they pass
+    """
+
+    # STEP 1: Fetch the rulegroup_to_events mapping for the project from redis
+    rulegroup_to_events = buffer.get_hash(model=Project, field={"project_id": project.id})
+
+    # STEP 2: Map each rule to the groups that must be checked for that rule.
+    rules_to_groups = get_rules_to_groups(rulegroup_to_events)
+
+    # STEP 3: Fetch the Rule models we need to check
+    alert_rules = Rule.objects.filter(id__in=list(rules_to_groups.keys()))
+
+    # STEP 4: Create a map of unique conditions to a tuple containing the JSON
+    # information needed to instantiate that condition class and the group_ids that
+    # must be checked for that condition. We don't query per rule condition because
+    # condition of the same class, interval, and environment can share a single scan.
+    condition_groups = get_condition_groups(alert_rules, rules_to_groups)
+
+    # Step 5: Instantiate each unique condition, and evaluate the relevant
+    # group_ids that apply for that condition
+    condition_group_results = get_condition_group_results(condition_groups, project, alert_rules[0])
```

Comment:
I'm just grabbing the first rule in the list here - it's required to have one to instantiate the class but nothing we're doing actually uses it, so I don't think it matters which one we use.

Reply:
Actually this metric uses it- https://github.com/getsentry/sentry/blob/master/src/sentry/rules/conditions/event_frequency.py#L212-L217 but since it's a _batch_ I'm not sure this metric makes sense to record anyway. It was added 4 years ago to track Snuba query usage when we introduced default alerts: https://github.com/getsentry/sentry/pull/17175

Reply:
Confirmed it's ok to remove this - I'll do so in a separate PR and rebase but this means I don't have to pass `rule` to   the condition instance at all ✨